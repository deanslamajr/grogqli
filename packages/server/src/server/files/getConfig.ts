import path from 'path';
import findUp from 'find-up';
import editJsonFile from 'edit-json-file';

const CONFIG_FILE_NAME = 'grogqli.json';

export interface GrogqliConfig {
  recordingsSaveDirectory: string;
  port: number;
  shouldDogFood: boolean;
}
const DEFAULT_CONFIG: GrogqliConfig = {
  recordingsSaveDirectory: path.join(process.cwd(), 'grogqli'),
  port: 4000,
  shouldDogFood: false,
};

type GetGrogqliConfigValue = (configKey: keyof GrogqliConfig) => string;
let getGrogqliConfigValue: GetGrogqliConfigValue | undefined;

export const getConfig = async (): Promise<GetGrogqliConfigValue> => {
  // return early if the response is cached
  if (getGrogqliConfigValue) {
    return getGrogqliConfigValue;
  }

  // recursively try to open a file with filename === CONFIG_FILE_NAME
  // if file doesn't exist, try again in parent directory
  // if file isn't found at root directory, return DEFAULT_CONFIG
  const configLocation = await findUp(
    async (directory) => {
      const hasConfig = await findUp.exists(
        path.join(directory, CONFIG_FILE_NAME)
      );
      return hasConfig ? directory : undefined;
    },
    { type: 'directory' }
  );

  let configFromFile: editJsonFile.JsonEditor;

  if (configLocation) {
    const configFileAbsolutePath = path.join(configLocation, CONFIG_FILE_NAME);
    configFromFile = editJsonFile(configFileAbsolutePath);
  }

  getGrogqliConfigValue = (configKey) => {
    if (configKey === undefined) {
      throw new Error(
        'Invalid use of getGrogqliConfigValue: must provide a key'
      );
    }
    let configValue: any;

    // First, try to resolve the value from the given config
    if (configLocation && configFromFile) {
      configValue = configFromFile.get(configKey);

      // Make 'recordingsSaveDirectory' an abs path
      if (
        configKey === 'recordingsSaveDirectory' &&
        configValue !== undefined
      ) {
        configValue = path.join(configLocation, configValue);
      }
    }

    // If config doesn't have a value, use the default
    if (configValue === undefined) {
      configValue = DEFAULT_CONFIG[configKey];
    }

    return configValue;
  };

  return getGrogqliConfigValue;
};
