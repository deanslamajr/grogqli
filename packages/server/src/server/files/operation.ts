import path from 'path';
import shortid from 'shortid';

import {
  OPERATIONS_FILENAME,
  OPERATIONS_FOLDER_NAME,
  getRecordingsRootDir,
  getSchemasFolderPath,
} from './';

export interface OperationFile {
  id: string;
  version: number;
  recordings: {
    [opRecordingId: string]: {
      id: string;
      rootTypeRecordings: {
        [rootTypeId: string]: {
          recordingId: string;
        };
      };
    };
  };
}

type CreateNewOpFile = (args: {
  opName: string;
  schemaId: string;
}) => OperationFile;

export const createNewOpFile: CreateNewOpFile = async (args) => {
  // - generate a new unique opId
  const opId = shortid.generate();
  // - add entry to grogqli/schemas/<opPlan.schemaId>/operations.json
  const opMapFile = await loadOperationsMappingFile(args.schemaId);

  // - create grogqli/operations/<opId>.json
  //   - recordings - empty object
};

type GetOpFileFromOpName = (params: {
  opName: string;
  schemaId: string;
}) => Promise<OperationFile | null>;

export const getOpFileFromOpName: GetOpFileFromOpName = async ({
  opName,
  schemaId,
}) => {
  const opsMappingFile = await loadOperationsMappingFile(schemaId);
  const opEntry = opsMappingFile.operations[opName];

  if (opEntry === undefined) {
    return null;
  }

  const operationFile = await getOperationFile(opEntry.id);

  if (operationFile === undefined) {
    return null;
  }

  return operationFile;
};

export const getOperationFile = async (
  operationId: string
): Promise<OperationFile | null> => {
  if (!operationId) {
    return null;
  }
  let operationFile: OperationFile;
  const recordingsRootDir = await getRecordingsRootDir();
  const pathToOperationFile = path.join(
    recordingsRootDir,
    OPERATIONS_FOLDER_NAME,
    `${operationId}.json`
  );

  try {
    operationFile = require(pathToOperationFile);
  } catch (error) {
    return null;
  }

  return operationFile;
};

export interface OperationsMappingFile {
  version: number;
  operations: {
    [opName: string]: {
      name: string;
      id: string;
    };
  };
}

export const loadOperationsMappingFile = async (
  schemaId: string
): Promise<OperationsMappingFile> => {
  const schemasFolderPath = await getSchemasFolderPath();
  // TODO after a feature is added that updates these `${schemaId}.json` files at runtime,
  // reevaluate whether or not this is necessary:
  // return JSON.parse(
  //   // not using require(`${schemaId}.json`) here (unlike in the resolvers)
  //   // bc we want to bypass the auto caching feature of require
  //   fs.readFileSync(path.join(schemaRecordingsPath, `${schemaId}.json`), 'utf8')
  // );
  const pathToOperationsData = path.join(
    schemasFolderPath,
    schemaId,
    OPERATIONS_FILENAME
  );
  let operationsData: OperationsMappingFile;
  try {
    operationsData = require(pathToOperationsData);
  } catch (error) {
    // TODO handle case where file doesnt exist for the given schemaId
    throw new Error(
      `TODO handle case where a operations file doesnt exist for the given schemaId. schemaId:${schemaId}`
    );
  }
  return operationsData;
};
