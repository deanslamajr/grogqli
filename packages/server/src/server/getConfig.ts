import path from 'path';
import findUp from 'find-up';

const CONFIG_FILE_NAME = 'grogqli.json';

interface GrogqliConfig {
  recordingsSaveDirectory: string;
  recordingsFilename: string;
}

const DEFAULT_CONFIG: GrogqliConfig = {
  recordingsSaveDirectory: '/home/dhan/codes/grogqli',
  recordingsFilename: 'groql_rec'
};
export const getConfig = async (): Promise<GrogqliConfig> => {
  console.log('process.cwd()', process.cwd());
  // TODO
  // recursively try to open a file with filename === CONFIG_FILE_NAME
  // if file doesn't exist, try again in parent directory
  // if file isn't found at root directory, return DEFAULT_CONFIG
  const configLocation = await findUp(CONFIG_FILE_NAME);

  console.log('configLocation', configLocation)

  // TODO
  // if a config doesn't have a value defined for a particular lookup
  // use the value found in DEFAULT_CONFIG
  return DEFAULT_CONFIG;
}