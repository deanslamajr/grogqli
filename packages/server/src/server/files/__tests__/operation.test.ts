import { loadOperationsMappingFile } from '../operation';
import { getConfig } from '../getConfig';
import path from 'path';

jest.mock('../getConfig');

describe('operation', () => {
  beforeEach(() => {
    const mockedGetConfig = getConfig as jest.MockedFunction<typeof getConfig>;
    mockedGetConfig.mockImplementation(async () => () => {
      return path.join(__dirname, 'grogqli');
    });
  });

  describe('getOperationsData', () => {
    it('should return the operations file associated with the given schemaId', async () => {
      const actual = await loadOperationsMappingFile('someSchemaId');
      expect(actual).toMatchSnapshot();
    });

    describe('if an operations file cannot be found that matches the given schemaId', () => {
      it('should throw', async () => {
        await expect(
          loadOperationsMappingFile('nonexistentSchemaId')
        ).rejects.toThrow();
      });
    });
  });
});
