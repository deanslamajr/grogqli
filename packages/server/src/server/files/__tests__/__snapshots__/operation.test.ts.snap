// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`operation getOperationsData should return the operations file associated with the given schemaId 1`] = `
Object {
  "operations": Object {
    "GetChores": Object {
      "id": "llK98u0",
      "name": "GetChores",
    },
    "someOpName": Object {
      "id": "someOpId",
      "name": "someOpName",
    },
  },
  "version": 1,
}
`;
